File: ejTasks/start_here/Pii/ObscuredPii.java
package Pii;

import java.util.function.UnaryOperator;

public class ObscuredPii<TValue> extends Pii<TValue> {
	public interface AltIdGenerator<TValue> {
		String obscure(TValue a_value);
	}
	
	private String m_altIdentifier;

    public ObscuredPii(TValue a_value, String a_altIdentifier) {
    	super(a_value);
        m_altIdentifier = a_altIdentifier;
    }

    public ObscuredPii(TValue a_value, AltIdGenerator<TValue> a_altIdentifierGenerator) {
    	super(a_value);
        m_altIdentifier = a_altIdentifierGenerator.obscure(a_value);
    }

    @Override
    public String toString() 
    {
        return m_altIdentifier;
    }

    public String toLoggable()
    {
        return m_altIdentifier;
    }

    @Override
    public Pii<TValue> cloneWithTransform(UnaryOperator<TValue> a_transform)
    {
        return new ObscuredPii<TValue>(a_transform.apply(exposeUnsecured()), m_altIdentifier);
    }
}

File: sharpTasks/start_here/pii/ObscuredPii.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace PiiLib
{
    public class ObscuredPii<TValue> : Pii<TValue>
    {
        public delegate string Obscurer(TValue a_value);

        private string m_altIdentifier;

        public ObscuredPii(TValue a_value, string a_altIdentifier)
            : base(a_value)
        {
            m_altIdentifier = a_altIdentifier;
        }

        public ObscuredPii(TValue a_value, Obscurer a_obscuringFunction)
            : base(a_value)
        {
            m_altIdentifier = a_obscuringFunction(a_value);
        }

        public override string ToString() 
        {
            return m_altIdentifier;
        }

        public string ToLoggable()
        {
            return m_altIdentifier;
        }

        public override Pii<TValue> CloneWithTransform(Func<TValue, TValue> a_transform)
        {
            return new ObscuredPii<TValue>(a_transform(ExposeUnsecured()), m_altIdentifier);
        }
    }
}
